packages/vite/src/node/optimizer/esbuildDepPlugin.ts

```diff
@@ -34,6 +34,21 @@ const externalTypes = [
   ...KNOWN_ASSET_TYPES
 ]
 
+const browserExternalContents = (id: string) => `\
+module.exports = new Proxy({}, {
+  get() {
+    return new Proxy({}, {
+      get() {
+        throw new Error(
+          'Module "${id}" has been externalized for ' +
+          'browser compatibility and cannot be accessed in client code.'
+        )
+      }
+    })
+  }
+})
+`
```

Comment:
So ESBuild looks into `contents` returned from `onLoads` before bundling it.

And since `browser-external:` modules currently are ES Modules with single default export, ESBuild bails upon seeing names re-export from such a module.

It's not possible make a module that exports any possible name. Hence a CommonJS module exporting a proxy.

But throwing immediately after a property is accessed is too early — it'll cause named import attempts fail at import time rather than usage time.

To avoid that, a `browser-external:` module now exports a proxy that for any property access returns a proxy that throws upon accessing its properties. 😵‍💫

Added a fix branch in the repro https://github.com/swandir/vite-browser-external-pnp/tree/fix

Reply:
Thanks for the explanation. I think this makes sense as I've seen vague errors in the past when code does `import { readFIleSync } from 'fs'` (named imports without yarn pnp). I haven't tested this so I'd have to take your word on it, or if there are test cases that would be great too.

I also think we should summarize your explanation in-code too.

Reply:
I'm afraid I can't add test cases here because they'd have to be run with Yarn PnP.

Without Yarn PnP this proxy module does not end up in the bundle.

Instead, for the example I linked in the previous comment, it does this:
```js
// (disabled):dep/node
var require_node = __commonJS({
  "(disabled):dep/node"() {
  }
});
```
<details>
<summary> __commonJS </summary>

```js
var __commonJS = (cb, mod) =>
  function __require() {
    return (
      mod || (0, cb[Object.keys(cb)[0]])((mod = { exports: {} }).exports, mod),
      mod.exports
    );
  };
```
</details>

And it's the same with or without my changes.

Interestingly, it does not produce an error on default import usage without PnP ([test](https://github.com/swandir/vite-browser-external-pnp/blob/36381ac06ed89a25ce0a9b81fe8dfb0687e47346/index.html#L20-L25)), as it's actually an empty object.

My version of the proxy module does throw, though: because any property of the outer proxy is a Proxy, and any property of the inner proxy explicitly throws.

Maybe it shouldn't though?

If I return empty `contents` from this `onLoad`, the output looks like this:
```js
// browser-external:./node
var require_node = __commonJS({
  "browser-external:./node"() {
  }
});
```
which is close to what ESBuild does. We might even change the namespace to `(disabled)`.

Reply:
I've been thinking that it'd be great to add Yarn PnP compatibility tests in [vite-ecosystem-ci](https://github.com/vitejs/vite-ecosystem-ci).

Would it be OK to add a repo for it in Vite's GitHub org?

Reply:
> I've seen vague errors in the past when code does import { readFIleSync } from 'fs' (named imports without yarn pnp)

This might be due to a change in ESBuild's begavior.

I guess earlier ESBuild did't handle such imports, and thus this onLoad was introduced in Vite.

Then at some point ESBuild just stopped calling this onLoad. Named import errors would stop appearing at that point.

Reply:
Oh, there're also subpath imports.

And named subpath imports produce runtime SyntaxError both with and without PnP.

```
Uncaught SyntaxError: The requested module '/@id/__vite-browser-external' does not provide an export named 'named'
```

Added a repro [here](https://github.com/swandir/vite-browser-external-pnp/blob/52eb2c72d794728fc7b9ea4daae690697e3d23ee/index.html#L5-L6).

Browser-disabled subpath imports are externalized here
https://github.com/vitejs/vite/blob/12ff29374085abac37a3aee0382482871b1710c5/packages/vite/src/node/plugins/resolve.ts#L828-L833

and are replaced here

https://github.com/vitejs/vite/blob/12ff29374085abac37a3aee0382482871b1710c5/packages/vite/src/node/plugins/resolve.ts#L276-L288

And this one is not fixable by switching to CommonJS as the proxy is not further transformed by ESBuild.

Reply:
Some for [builtins](https://github.com/swandir/vite-browser-external-pnp/blob/7abfdea019cf3510b3ed4c4e825ac0aedea51a2a/index.html#L7-L8).

Reply:
I think this would make the most sense right now:
- return empty contents for disabled modules in esbuildDepPlugin as it's more aligned with what's happening without PnP
- maybe register this onLoad only in PnP mode to make the purpose more explicit
- do nothing about named subpath and builtin imports, since it does not make much sense to import these directly in browser-targeting code

@bluwy sounds good?

Reply:
I'm confused, but probably due to my lack of understanding of yarn pnp. I don't quite understand how the proxy and pnp are related. The proxy is needed to provide user-friendly error messages on the client side. I don't think we should remove it.

> I've been thinking that it'd be great to add Yarn PnP compatibility tests in [vite-ecosystem-ci](https://github.com/vitejs/vite-ecosystem-ci).
> 
> Would it be OK to add a repo for it in Vite's GitHub org?

That would be a great addition, though I'm not sure if the setup supports pnp at the moment.

Reply:
> That would be a great addition, though I'm not sure if the setup supports pnp at the moment.

Looks like it can detect Yarn in a repo.

If scripts are run via Yarn in an external repo, that'll be enough to enable PnP.

I'll try it out later.